src/mcts/search.h

```diff
@@ -236,17 +238,26 @@ class SearchWorker {
     Node* node;
     bool is_collision = false;
     bool nn_queried = false;
+    bool is_cache_hit = false;
```

Comment:
With 3 bools in a row, do modern compilers optimize these to fit all into one byte? Or do alignment issues mean that such an optimization wouldn't help anyways?

Reply:
I tried it by adding "/d1reportSingleClassLayoutNode" as a compiler option. It result in a ton of information on the output but at the top is this: 

: added class lczero::Node	size(72):
2>	+---
2> 0	| EdgeList edges_
2>16	| index_
2>  	| <alignment member> (size=2)
2>20	| q_
2>24	| n_
2>28	| n_in_flight_
2>  	| <alignment member> (size=2)
2>32	| visited_policy_
2>36	| max_depth_
2>38	| full_depth_
2>40	| is_terminal_
2>  	| <alignment member> (size=7)
2>48	| parent_
2>56	| ?$unique_ptr@VNode@lczero@@U?$default_delete@VNode@lczero@@@std@@ child_
2>64	| ?$unique_ptr@VNode@lczero@@U?$default_delete@VNode@lczero@@@std@@ sibling_
2>	+---

Reply:
wow, that bool is a killer for alignment purposes. and it's a bit odd that it doesn't combine `index_` and `n_in_flight_` ? Does it preserve the order in which they're declared?

Reply:
Note also by adding "#pragma pack(1)" above the class definition the size is reduced. This is not always advisable thought as mis-alignment can have other performance problems.

class lczero::Node	size(61):
1>	+---
1> 0	| EdgeList edges_
1>16	| index_
1>18	| q_
1>22	| n_
1>26	| n_in_flight_
1>28	| visited_policy_
1>32	| max_depth_
1>34	| full_depth_
1>36	| is_terminal_
1>37	| parent_
1>45	| ?$unique_ptr@VNode@lczero@@U?$default_delete@VNode@lczero@@@std@@ child_
1>53	| ?$unique_ptr@VNode@lczero@@U?$default_delete@VNode@lczero@@@std@@ sibling_
1>	+---

Reply:
@dubslow I lurking and trying to observe what is important and what I can contribute, if anything. I gather that the size of Node is important. You can reduce easily by simple moving the declaration of n_in_flight to right before index. Node is now smaller.

1>class lczero::Node	size(64):
1>	+---
1> 0	| EdgeList edges_
1>16	| index_
1>18	| n_in_flight_
1>20	| q_
1>24	| n_
1>28	| visited_policy_
1>32	| max_depth_
1>34	| full_depth_
1>36	| is_terminal_
1>  	| <alignment member> (size=3)
1>40	| parent_
1>48	| ?$unique_ptr@VNode@lczero@@U?$default_delete@VNode@lczero@@@std@@ child_
1>56	| ?$unique_ptr@VNode@lczero@@U?$default_delete@VNode@lczero@@@std@@ sibling_

Reply:
Indeed :) size of Node is important, though less important than it once was, since we have Edge-Node separation now. Each node reported by the search spawns one `Node`, but also all the relevant child `Edges` for that node, so if the search tree has n nodes reported, then on average the search tree takes `n * (sizeof(Node) + 35 * sizeof(Edge))`, where 35 is the average chess branching factor. So the size of Edges dominates (see PR176), but at the same time we could probably get a several-percent improvement here with the alignment. Indeed rearranging the 2 byte ints helps, but I think if we're going to rearrange for optimization, we may as well go whole-hog and put all members in order of size, i.e. pointers first. That way it's easier to modify in the future. I'll make another PR shortly. Many thanks for the pointer to the correct compiler flag, and of course printing the output here to spike my curiosity!

Reply:
Given this disparity, are we sure edges don't have any similar problems?

Reply:
@dubslow I arranged things as you suggested and it comes out nicely:

1>class lczero::Node	size(64):
1>	+---
1> 0	| EdgeList edges_
1>16	| ?$unique_ptr@VNode@lczero@@U?$default_delete@VNode@lczero@@@std@@ child_
1>24	| ?$unique_ptr@VNode@lczero@@U?$default_delete@VNode@lczero@@@std@@ sibling_
1>32	| parent_
1>40	| q_
1>44	| n_
1>48	| visited_policy_
1>52	| index_
1>54	| n_in_flight_
1>56	| max_depth_
1>58	| full_depth_
1>60	| is_terminal_
1>  	| <alignment member> (size=3)

Reply:
Yes. They are currently 6 bytes theoretically, 8 in practice on my g++; PR176 as mentioned will reduce that to 4 and 4 respectively. Edges have only two members.

Reply:
Turns out the EdgeList itself has 6 bytes of padding. I guess that would be next on the list.

Reply:
@oscardssmith This is the result for the Edge class

1>class lczero::Edge	size(8):
1>	+---
1> 0	| Move move_
1>  	| <alignment member> (size=2)
1> 4	| p_
1>	+---

Moving p_ to be first it's a tiny bit better:

1>class lczero::Edge	size(8):
1>	+---
1> 0	| p_
1> 4	| Move move_
1>  	| <alignment member> (size=2)
1>	+---

I doubt making this change would have any measurable impact.

Reply:
Three bools in a row won't share the same byte (C++ requires that it should be possible to take address of each).
However, i can write:
```cpp
    bool is_collision:1 = false;
    bool nn_queried:1 = false;
    bool is_cache_hit:1 = false;
```
and then they will indeed reside in one byte.

But it probably will slow down things.  NodeToProcess is not memory critical structure at all, there are only minibatch_size * num_threads of those in total at any moment.
So even if we'll ever be able to optimize code to allow 64 threads and minibatch of 2000 nodes in each (unrealistic), it will still be <2MB in total.

Reply:
We happy engineers who live in the days when 2MB can be considered trivial :)

Reply:
Well if we have 64 threads and 2000 nodes in each, and every thread does say 4 NN queries per second, that's 500000nps, which is memory growth of ~100MB per second.
Comparing to that, static 2MB is fine.

Current reality is more like 2 threads of 256 nodes, which is more like 5KB.